modules/carto/src/layers/heatmap.ts

```diff
@@ -0,0 +1,181 @@
+import type {ShaderPass} from '@luma.gl/shadertools';
+import {random} from '@luma.gl/shadertools';
+import {colorCategories} from '@deck.gl/carto';
+const glsl = (s: TemplateStringsArray) => `${s}`;
+
+/**
+ * @filter       Heatmap
+ * @param radius Blur radius, controls smoothness of heatmap
+ * @param rangeScale Scale factor to apply to values prior to applying color scale
+ * @param color1-6 Colors to use in color scale
+ */
+
+const fs = glsl`\
+uniform heatmapUniforms {
+  float radius;
+  float rangeScale;
+  vec3 color1;
+  vec3 color2;
+  vec3 color3;
+  vec3 color4;
+  vec3 color5;
+  vec3 color6;
+} heatmap;
+
+// Controls quality of heatmap, larger values increase quality at expense of performance
+const float SUPPORT = 8.0;
+const vec4 STOPS = vec4(0.2, 0.4, 0.6, 0.8);
+
+vec3 colorGradient(float value) {
+  vec3 c1;
+  vec3 c2;
+  vec2 range;
+  if (value < STOPS.x) {
+    range = vec2(0.0, STOPS.x);
+    c1 = heatmap.color1; c2 = heatmap.color2;
+  } else if (value < STOPS.y) {
+    range = STOPS.xy;
+    c1 = heatmap.color2; c2 = heatmap.color3;
+  } else if (value < STOPS.z) {
+    range = STOPS.yz;
+    c1 = heatmap.color3; c2 = heatmap.color4;
+  } else if (value < STOPS.w) {
+    range = STOPS.zw;
+    c1 = heatmap.color4; c2 = heatmap.color5;
+  } else if (value < 1.0 ) {
+    range = vec2(STOPS.w, 1.0);
+    c1 = heatmap.color5; c2 = heatmap.color6;
+  } else {
+    // Fade out to white
+    range = vec2(1.0, 10.0);
+    c1 = heatmap.color6; c2 = vec3(1.0);
+  }
+
+  float f = (value - range.x) / (range.y - range.x);
+  return mix(c1, c2, f);
+}
+
+vec4 heatmap_sampleColor(sampler2D source, vec2 texSize, vec2 texCoord) {
+  float accumulator = 0.0;
+  float total = 0.0;
+
+  /* randomize the lookup values to hide the fixed number of samples */
+  float offset = random(vec3(12.9898, 78.233, 151.7182), 0.0);
```

Comment:
> randomize the lookup values to hide the fixed number of samples

If we needed to reduce the grain further without increasing samples, perhaps we could set a random seed for each frame, and accumulate samples to the framebuffer, so that the noise stabilizes to a smooth image within a few frames when the map is not in motion.

Strictly an idea for future versions though. I like the texture of the random noise quite a lot. :)

Reply:
Yes, that would be nice - although you might get a motion-blur effect if you move the map too fast

Reply:
Yes, would need to clear the FBO if view state has changed. I did the same thing here at one point...

https://threejs.org/examples/?q=hash#webgl_materials_alphahash

... which can be enabled/disabled with the "temporal anti-aliasing" toggle.

In any case, no changes needed IMHO!

Reply:
I like this idea, but will not implement it in this PR. It would be an interesting addition to the whole RTT pattern.

From my tests, the underlying `QuadbinTileLayer` is the perf bottleneck, due to it using a `SolidPolygonLayer`, when it could do something much more lean, like the `RasterLayer`